src/main/scala/net/ripe/rpki/publicationserver/store/postgresql/PgStore.scala

```diff
@@ -170,6 +168,24 @@ class PgStore(val pgConfig: PgConfig) extends Hashing with Logging {
       // (which most often is the CA owning the objects)
       sql"SELECT acquire_client_id_lock(${clientId.value})".execute().apply()
 
+      changeSet.pdus.foreach {
+        case PublishQ(uri, _, None, _) =>
```

Comment:
Why do we do it before saving objects and together with saving objects?

Reply:
Before saving objects we check to prohibit messages like:

```
<withdraw uri="foo" hash="abcd"/>
<publish uri="foo">XYZXYZ</publish>
```

Which should be a single `<publish uri="foo" hash="abcd">XYZXYZ</publish>`.

The check during the save is done to disallow:

```
<publish uri="foo">XYZXYZ</publish>
<publish uri="foo">XYZXYZ</publish>
```

We could also check that each `uri` in a message is unique, but I found that harder to translate to an error code as defined in the RFC. So checking both stages will give a reasonable error message (object does not exist, hash does not match, object already exists).